[Law Sean Meng] iP

seox123 opened the pull request:
## Doemon
> "Gadgets are only useful if they are used wisely." – Doraemon ([Source](https://www.upgradingoneself.com/doraemon-quotes/))

Doemon frees your brain from the burden of existing! With Doemon's [Memory Bread](https://doraemon.fandom.com/wiki/Copying_Toast#:~:text=%22Memorization%20Bread%22\)%20or%20Memory,contents%20'printed'%20on%20it.) ~and his name~ that he stole from Doraemon, you can now:
* **STOP** thinking
* **START** recording
* profit

So you may be wondering...How do I get my filthy ✋✋ all over this delicious piece of 🍞?
Just follow a few simple steps silly 🤪:
1. Download my latest release [here](https://github.com/seox123/ip/releases/tag/A-Jar).
2. Place it inside an empty folder/directory
3. Double click it to run 🏃!

Moldy, moldy features:
- [x] Managing tasks
- [ ] Talk to me (coming soon!)
- [ ] Eat me (coming soon!)

___

Here's a little bread crumb for you to ```eat```: 
```java
public class Bread {
    public static void main(String args[]) {
        Bread.eat();
    }
    
    private static void eat() {
        System.out.println("Crumb eaten!");
    }
}
```

Commit:
Add Gradle support

seox123 committed:
Changed logo and name

seox123 committed:
Added echo functionality and exit on 'bye' command

seox123 committed:
Formatted the output by adding indentation and separating lines

seox123 committed:
Renamed the bot and related files, and shifted some parts of the main function out

seox123 committed:
Added add task functionality

seox123 committed:
Added list tasks functionality

seox123 committed:
Created the Task class and changed the string representation of tasks into Task objects

seox123 committed:
Added mark/unmark functionality to added tasks

seox123 committed:
Added Todo, Deadlines, and Events

seox123 committed:
Changed doemon logo and added some test cases

seox123 committed:
Added some exception handling and added test cases for the new exceptions

seox123 committed:
Changed tasks array to an ArrayList

seox123 committed:
Added delete task functionality, added new InvalidTaskNumberException, and added new test cases to test the new functions

seox123 committed:
Added load task functionality from txt file

seox123 committed:
Added save task functionality

heyzec commented on src/main/java/Doemon.java:
```diff
@@ -0,0 +1,189 @@
+    /**
+     * Picture of Doemon made using text
+     */
+    private static String logo = "                       _______________\n" +
+            "                      /  --. --.      \\ \n" +
+            "                     /  | '| ' |   \\   \\ \n" +
+            "                    / /  `-O--'     \\   \\ \n" +
+            "                   |.  --  |  --     |   |\n" +
+            "                   |  --   |  --     |   |\n" +
+            "                    \\  (___|_______) /  /\n" +
+            "                     \\              /  /\n" +
+            "                       |== (t) ===|____";
```
Nice art! It can be made final and a constant though!

heyzec commented on src/main/java/Doemon.java:
```diff
@@ -0,0 +1,189 @@
+     * Starts the Doemon chatbot.
+     */
+    private void start() {
+        // Handling inputs
+        Scanner sc = new Scanner(System.in);
```
Perhaps this can be migrated into the Ui class.

heyzec commented on src/main/java/Doemon.java:
```diff
@@ -0,0 +1,189 @@
+                int listNum = 1;
+                String listStr = "Here is what's on my bread:\n\t";
+                for (Task task : tasks) {
+                    if (task == null) break;
```
Our coding standards doesn't like if-else without braces.

heyzec commented on src/main/java/Doemon.java:
```diff
@@ -0,0 +1,189 @@
+            // Add task
+            try {
+                addTask(inputStr, inputArr);
+            } catch (EmptyTaskException ete) {
+                System.out.println(output(ete.toString()));
+            } catch (InvalidTaskException ite) {
+                System.out.println(output(ite.toString()));
+            } catch (MissingArgumentException mae) {
+                System.out.println(output(mae.toString()));
```
If I'm not wrong, it is possible to catch multiple exceptions under the same clause.

heyzec commented on src/main/java/Doemon.java:
```diff
@@ -0,0 +1,189 @@
+                && (inputArr[0].equals("mark") || inputArr[0].equals("unmark") || inputArr[0].equals("delete"))
+                && isInteger(inputArr[1])) {
+            int index = Integer.parseInt(inputArr[1]) - 1;
+            if (index >= 0 && index < this.tasks.size()) {
+                if (inputArr[0].equals("mark")) {
+                    this.tasks.get(index).mark();
+                    System.out.println(
+                            output(String.format("Yay! This task is now marked as done:\n\t  %s",
+                                    this.tasks.get(index).toString())));
+                } else if (inputArr[0].equals("unmark")){
+                    this.tasks.get(index).unmark();
+                    System.out.println(
+                            output(String.format("I guess you weren't done with that one:\n\t  %s",
+                                    this.tasks.get(index).toString())));
+                } else {
+                    Task removed = this.tasks.remove(index);
+                    System.out.println(
+                            output(String.format("I used a knife to slice off this task from my bread:\n\t  %s" +
+                                            "\n\tThere are %d items left on my bread.",
+                                    removed.toString(),
+                                    this.tasks.size())));
+                }
+                return true;
+            } else {
+                throw new InvalidTaskNumberException(inputArr[0]);
+            }
```
I would suggest converting this into a guard clause style.

heyzec commented on src/main/java/Doemon.java:
```diff
@@ -0,0 +1,189 @@
+                throw new InvalidTaskNumberException(inputArr[0]);
+            }
+        }
+        return false;
```
Can also be converted to a guard clause.

heyzec commented on src/main/java/Doemon.java:
```diff
@@ -0,0 +1,189 @@
+    private void addTask(String inputStr, String[] inputArr) throws EmptyTaskException, InvalidTaskException,
+            MissingArgumentException {
+        // Add item to list of tasks
+        if (inputArr[0].equals("todo")) {
+            String detail = inputStr.substring(4).trim();
+            if (detail.trim().equals("")) throw new EmptyTaskException("todo");
+            this.tasks.add(new Todo(detail));
+        } else if (inputArr[0].equals("deadline")) {
+            String[] details = inputStr.substring(8).trim().split(" /by ");
+            if (details[0].trim().equals("")) throw new EmptyTaskException("deadline");
+            if (details.length == 1) throw new MissingArgumentException("deadline", "/by");
+            this.tasks.add(new Deadline(details[0], details[1]));
+        } else if (inputArr[0].equals("event")) {
+            String[] details = inputStr.substring(5).trim().split(" /at ");
+            if (details[0].trim().equals("")) throw new EmptyTaskException("event");
+            if (details.length == 1) throw new MissingArgumentException("event", "/at");
+            this.tasks.add(new Event(details[0], details[1]));
+        } else {
+            throw new InvalidTaskException();
+        }
```
Can consider converting this to a switch to avoid the repeated `.equals()`

heyzec reviewed (commented):
Nice UI text responses by the Doemon bot!

florentianayuwono commented on src/main/java/Doemon.java:
```diff
@@ -0,0 +1,189 @@
+    /**
+     * Picture of Doemon made using text
+     */
+    private static String logo = "                       _______________\n" +
```
Nice! Perhaps it can be written as `private static final String LOGO`?

florentianayuwono commented on src/main/java/Doemon.java:
```diff
@@ -0,0 +1,189 @@
+     * @return a boolean indicating if the command was a mark/unmark/delete command
+     * @throws InvalidTaskNumberException if the task number indicated does not exist
+     */
+    private boolean hasMarkedOrDeleted(String[] inputArr) throws InvalidTaskNumberException{
```
Perhaps it's good to shorten the method length by creating separate methods to handle the `System.out.println` and parsing commands? So that it can follow this coding standard:

> Avoid Long Methods
  Can improve code quality using technique: avoid long methods
Be wary when a method is longer than the computer screen, and take corrective action when it goes beyond 30 LOC (lines of code). The bigger the haystack, the harder it is to find a needle.

florentianayuwono commented on src/main/java/MissingArgumentException.java:
```diff
@@ -0,0 +1,33 @@
+public class MissingArgumentException extends DoemonException {
```
I like how you create various child classes for `DaemonException`.👍

florentianayuwono commented on src/main/java/Todo.java:
```diff
@@ -0,0 +1,18 @@
+    /**
+     * Returns a string representation of this object.
+     * @return a string representing this object
+     */
```
I like how you have JavaDoc for almost (if not all) methods and fields.

florentianayuwono reviewed (commented):
Overall, I think you did a good job with the chatbot. I would suggest to continue implementing some OOP as recommended in the Week 3 iP to shorten some methods or classes. Great work!🌞

yixiann commented on src/main/java/Doemon.java:
```diff
@@ -0,0 +1,189 @@
+                String listStr = "Here is what's on my bread:\n\t";
+                for (Task task : tasks) {
+                    if (task == null) break;
+                    listStr += listNum++ + "." + task.toString() + "\n\t";
```
Concatenating strings may not be very efficient. Maybe you can consider using StringBuilder instead? 

```
if (inputStr.equals("list")) {
    int listNum = 1;
    StringBuilder listStr = new StringBuilder("Here is what's on my bread:\n\t");
    for (Task task : tasks) {
        if (task == null) break;
        listStr.append(listNum++).append(".").append(task).append("\n\t");
    }
    listStr = new StringBuilder(listNum == 1 ? "You have no tasks!" : listStr.toString().trim());
    System.out.println(output(listStr.toString()));
    continue;
}
```

yixiann reviewed (commented):
Generally good! Appropriate naming of variables.

seox123 committed:
Add ability to read date/time

seox123 committed:
Merge branch 'branch-Level-8'

seox123 committed:
Fix small bug with save string of deadline and task

seox123 committed:
Refactor code to extract closely related code as classes

seox123 committed:
Refactor related code into separate packages

seox123 committed:
Fix runtest.bat file

seox123 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

seox123 committed:
Add gradle support

seox123 committed:
Fix gradle error

seox123 committed:
Add some JUnit test cases for task package

seox123 committed:
Edit build.gradle file to create v0.1 jar

seox123 committed:
Add JavaDoc comments

seox123 committed:
Tweak code to comply with coding standard

seox123 committed:
Add find command functionality

seox123 committed:
Add JavaDocs for Ui class

seox123 committed:
Tweak code in Ui class to comply with coding standard

seox123 committed:
Change find command to search using displayed task string

seox123 committed:
Merge branch 'branch-A-JavaDoc'

seox123 committed:
Merge branch 'branch-A-CodingStandard'

seox123 committed:
Merge branch 'branch-Level-9'

seox123 committed:
Add checkstyle support

seox123 committed:
Create GUI for Doemon chat bot using javafx and FXML

seox123 committed:
Merge branch 'branch-A-CheckStyle'

seox123 committed:
Merge branch 'branch-Level-10'

seox123 committed:
Add assert statements

seox123 committed:
Fix checkstyle errors and add javadocs for GUI files

seox123 committed:
Parser class: refactor code within parse function

The parse() method is long.

The length of the code in the parse() method negatively impacts
readability, making it difficult to identify the logical flow of
the code. This increases the possibility of errors in the code
going unnoticed.

Refactoring code into smaller, more focused methods allow us to
know where to look when we decide to change certain parts of how
the Parser class checks for different commands.

Let's refactor the command-checking logic into their own methods.

Each command requires multiple lines of code to check for validity.
Refactoring the code into their own methods declutters the parse()
method, and allows for easier finding and changing of checking logic
for the respective commands in the future.

seox123 committed:
Storage class: Flatten arrow code and reduce nesting

seox123 committed:
Minor change to JavaDocs in Response class

seox123 committed:
Merge pull request #2 from seox123/branch-A-Assertions

Add assertions

seox123 committed:
Merge branch 'master' into branch-A-CodeQuality

seox123 committed:
Merge pull request #3 from seox123/branch-A-CodeQuality

Improve code quality

seox123 committed:
Add a gradle.yml file for continuous integration

seox123 committed:
Add help command

seox123 committed:
Fix error in parse logic

seox123 committed:
Merge branch 'branch-C-Help'

seox123 committed:
Add shortform commands

seox123 committed:
Merge branch 'branch-C-FriendlierSyntax'

seox123 committed:
Fix deadline and event commands

seox123 committed:
Improve GUI and response message display

seox123 committed:
Move GUI background into images folder

seox123 committed:
Add user guide, tweak some response strings, and add shortform for find command

seox123 committed:
Remove all unnecessary usages of the 'this' keyword

seox123 committed:
Change bye command to close the window

seox123 committed:
Fix assert error in Task constructor method

seox123 committed:
Remove text ui tests from gradle.yml file
